### PR TITLE
[CU-338n06m] versioning working

### PR DESCRIPTION
**Pull request by lucasazuma**
N versions working

https://user-images.githubusercontent.com/56770560/194683152-57ef89f6-00b9-49e2-ac85-30431e0d6cc2.mp4

### CODE REVIEW

File: lib/generators/rest_api_generator/helpers.rb [hunk 1]
```diff
@@ -17,5 +17,14 @@ def editable_attributes
...
     end
+
+    def versioning?
+      !options["scope"].empty?
+    end
+
+    def version
+      parts = options["scope"].split(".")
+      "module " + parts[0].capitalize + '::' + parts[1].capitalize
+    end
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/helpers.rb:28]: This works for only two version? Can we change this to work with n versions?

File: lib/generators/rest_api_generator/helpers.rb [hunk 2]
```diff
@@ -17,5 +17,14 @@ def editable_attributes
         Rails::Generators::GeneratedAttribute.new(column.name.to_s, column.type.to_s)
       end
     end
+
+    def versioning?
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/helpers.rb:21]: I dont know if versiong is the best name for this feature, we would like to be like a custom path for the model and controller that can be used for versioning, maybe stay with name scope is better

File: lib/generators/rest_api_generator/templates/rest_api_controller.rb.tt [hunk 3]
```diff
@@ -1,3 +1,6 @@
+<% if versioning? -%>
+<%= version %>
+<% end -%>
 class <%= class_name.pluralize %>Controller < ApplicationController
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/templates/rest_api_controller.rb.tt:4]: Here we need indentation for the class, rails uses a function called `model_namespacing` for this feature, i think we should follow them

File: lib/generators/rest_api_generator/resource_generator.rb [hunk 4]
```diff
@@ -40,5 +40,15 @@ def build_model_attributes
       end
       model_attributes
     end
+
+    def define_scope
+      if options["scope"].empty?
+        API_CONTROLLER_DIR_PATH
+      else
+        parts = options["scope"].split(".")
+        API_CONTROLLER_DIR_PATH + "/#{parts[0]}/#{parts[1]}"
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/resource_generator.rb:49]: Same thing from version method, can we change to work with n scopes?

File: lib/generators/rest_api_generator/resource_generator.rb [hunk 5]
```diff
@@ -15,20 +15,20 @@ class ResourceGenerator < Rails::Generators::NamedBase
...
-        template "child_api_controller.rb", controller_path
-        template "child_api_spec.rb", controller_test_path
-      end
+      # else
+      #   scope_path = options["scope"].pluralize
+      #   controller_path = "#{API_CONTROLLER_DIR_PATH}/#{scope_path}/#{file_name.pluralize}_controller.rb"
+      #   controller_test_path = "#{API_TEST_DIR_PATH}/#{scope_path}/#{file_name.pluralize}_controller_spec.rb"
+      #   template "child_api_controller.rb", controller_path
+      #   template "child_api_spec.rb", controller_test_path
+      # end
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/resource_generator.rb:31]: Children options was removed for now, alright? Can you just delete this code for now? In next features we will add this back

**Review by PedroAugustoRamalhoDuarte — CHANGES_REQUESTED** [review]
@lucasazuma Amazing feature, can you remove rest-api-generator-0.1.0.gem too, this cannot go to git

File: lib/generators/rest_api_generator/helpers.rb (hunk 1 above)
Comment (lucasazuma) [reply lib/generators/rest_api_generator/helpers.rb:28]: done, now it works with n versions

File: lib/generators/rest_api_generator/resource_generator.rb (hunk 4 above)
Comment (lucasazuma) [reply lib/generators/rest_api_generator/resource_generator.rb:49]: done, now it works with n versions

File: lib/generators/rest_api_generator/helpers.rb (hunk 2 above)
Comment (lucasazuma) [reply lib/generators/rest_api_generator/helpers.rb:21]: done

File: lib/generators/rest_api_generator/resource_generator.rb [hunk 6]
```diff
@@ -15,20 +15,12 @@ class ResourceGenerator < Rails::Generators::NamedBase
...
-        route routes_string
-      else
-        scope_path = options["scope"].pluralize
-        controller_path = "#{API_CONTROLLER_DIR_PATH}/#{scope_path}/#{file_name.pluralize}_controller.rb"
-        controller_test_path = "#{API_TEST_DIR_PATH}/#{scope_path}/#{file_name.pluralize}_controller_spec.rb"
-        template "child_api_controller.rb", controller_path
-        template "child_api_spec.rb", controller_test_path
-      end
+      controller_path = "#{define_scope}/#{file_name.pluralize}_controller.rb"
+      controller_test_path = "#{API_TEST_DIR_PATH}/#{file_name.pluralize}_controller_spec.rb"
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/resource_generator.rb:19]: Teste file needs to be scoped too

File: lib/generators/rest_api_generator/templates/scope_rest_api_controller.rb.tt [hunk 7]
```diff
@@ -0,0 +1,39 @@
+
+<%= version %>
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/templates/scope_rest_api_controller.rb.tt:2]: I think we can get this working with only one file with `module_namespacing` from rails:
https://github.com/rspec/rspec-rails/blob/main/lib/generators/rspec/controller/templates/controller_spec.rb
https://apidock.com/rails/v4.2.7/Rails/Generators/NamedBase/module_namespacing

File: lib/generators/rest_api_generator/helpers.rb [hunk 8]
```diff
@@ -17,5 +17,18 @@ def editable_attributes
         Rails::Generators::GeneratedAttribute.new(column.name.to_s, column.type.to_s)
       end
     end
+
+    def scope?
+      !options["scope"].empty?
+    end
+
+    def version
```
Comment (PedroAugustoRamalhoDuarte) [inline lib/generators/rest_api_generator/helpers.rb:25]: I dont know with version is the best name for this functions, maybe namespace?

**Review by PedroAugustoRamalhoDuarte — CHANGES_REQUESTED** [review]
And we need to delete this file from commits: [rest-api-generator-0.1.0.gem](https://github.com/SwitchDreams/rest-api-generator/pull/4/files#diff-0579a42b42af8df6b8eba749c369999971859dac21d715dbe13b45d5872f3822)

**Comment by waliffcordeiro** [comment]
@lucasazuma when you finish a conversation, click on 'resolve conversation' to make it easier to see the PR

**Review by PedroAugustoRamalhoDuarte — COMMENTED** [review]
I will review this in the afternoon

**Pull request merged**